File: scripts/refactor.mjs
import fs from 'fs';

function refactorFile(filePath) {
  let content = fs.readFileSync(filePath, 'utf8');

  // Add Turso import if not exists
  if (!content.includes("import { turso } from './turso';")) {
    content = content.replace(/import Database from 'better-sqlite3';/, "import { turso } from './turso';");
  }
  
  // Replace function definitions with async
  content = content.replace(/export function ([a-zA-Z0-9_]+)\(/g, 'export async function $1(');
  
  // Replace the inside of the functions
  // Instead of complex regex, we just replace the db.prepare
  
  content = content.replace(/const db = get[a-zA-Z0-9_]+DB\(\);/g, '');
  content = content.replace(/db\.prepare\((.*?)\)\.get\((.*?)\)/g, '(await turso.execute({ sql: $1, args: [$2] })).rows[0]');
  content = content.replace(/db\.prepare\((.*?)\)\.get\(\)/g, '(await turso.execute($1)).rows[0]');
  
  content = content.replace(/db\.prepare\((.*?)\)\.all\((.*?)\)/g, '(await turso.execute({ sql: $1, args: [$2] })).rows');
  content = content.replace(/db\.prepare\((.*?)\)\.all\(\)/g, '(await turso.execute($1)).rows');
  
  content = content.replace(/db\.prepare\((.*?)\)\.run\((.*?)\)/g, 'await turso.execute({ sql: $1, args: [$2] })');
  content = content.replace(/db\.prepare\((.*?)\)\.run\(\)/g, 'await turso.execute($1)');

  fs.writeFileSync(filePath, content, 'utf8');
}

const libs = ['src/lib/player-db.ts', 'src/lib/room-db.ts', 'src/lib/music-db.ts'];
libs.forEach(f => {
  if (fs.existsSync(f)) {
    refactorFile(f);
  }
});

File: scripts/refactor.py
import re

with open('src/lib/room-db.ts', 'r') as f:
    text = f.read()

# Replace better-sqlite3 imports
text = re.sub(r"import Database from 'better-sqlite3';\n?", "import { turso } from './turso';\n", text)
text = re.sub(r"let dbInstance.*?\n", "", text)
text = re.sub(r"export function getRoomDB.*?\{.*?\}", "", text, flags=re.DOTALL)
text = re.sub(r"const db = getRoomDB\(\);\n?", "", text)

# Fix signatures
def async_sig(match):
    name = match.group(1)
    args = match.group(2)
    ret = match.group(3)
    if "Promise<" not in ret:
        ret = f"Promise<{ret}>"
    return f"export async function {name}({args}): {ret} {{"

text = re.sub(r"export function (\w+)\((.*?)\): (.*?) \{", async_sig, text)

# Convert all stmt.run(...) or db.prepare(...).run/get/all
# Since python regex is easier:

# db.prepare(SQL).get(ARGS)
text = re.sub(r"db\.prepare\(`([\s\S]*?)`\)\.get\(([\s\S]*?)\)", r"(await turso.execute({ sql: `\1`, args: [\2] })).rows[0] as any", text)
text = re.sub(r"db\.prepare\('(.*?)'\)\.get\(([\s\S]*?)\)", r"(await turso.execute({ sql: '\1', args: [\2] })).rows[0] as any", text)

# db.prepare(SQL).all(ARGS)
text = re.sub(r"db\.prepare\(`([\s\S]*?)`\)\.all\(([\s\S]*?)\)", r"(await turso.execute({ sql: `\1`, args: [\2] })).rows as any", text)
text = re.sub(r"db\.prepare\('(.*?)'\)\.all\(([\s\S]*?)\)", r"(await turso.execute({ sql: '\1', args: [\2] })).rows as any", text)

# db.prepare(SQL).run(ARGS)
text = re.sub(r"db\.prepare\(`([\s\S]*?)`\)\.run\(([\s\S]*?)\)", r"await turso.execute({ sql: `\1`, args: [\2] })", text)
text = re.sub(r"db\.prepare\('(.*?)'\)\.run\(([\s\S]*?)\)", r"await turso.execute({ sql: '\1', args: [\2] })", text)

# db.prepare(SQL).get()
text = re.sub(r"db\.prepare\(`([\s\S]*?)`\)\.get\(\)", r"(await turso.execute(`\1`)).rows[0] as any", text)
text = re.sub(r"db\.prepare\('(.*?)'\)\.get\(\)", r"(await turso.execute('\1')).rows[0] as any", text)
text = re.sub(r'db\.prepare\("(.*?)"\)\.get\(\)', r'(await turso.execute("\1")).rows[0] as any', text)

# db.prepare(SQL).all()
text = re.sub(r"db\.prepare\(`([\s\S]*?)`\)\.all\(\)", r"(await turso.execute(`\1`)).rows as any", text)
text = re.sub(r"db\.prepare\('(.*?)'\)\.all\(\)", r"(await turso.execute('\1')).rows as any", text)
text = re.sub(r'db\.prepare\("(.*?)"\)\.all\(\)', r'(await turso.execute("\1")).rows as any', text)

# db.prepare(SQL).run()
text = re.sub(r"db\.prepare\(`([\s\S]*?)`\)\.run\(\)", r"await turso.execute(`\1`)", text)
text = re.sub(r"db\.prepare\('(.*?)'\)\.run\(\)", r"await turso.execute('\1')", text)

# const stmt = db.prepare(SQL) ... stmt.run(ARGS)
# Because it's hard with regex, let's just do it precisely for room-db.ts
text = text.replace("const stmt = db.prepare(`", "/* stmt */ const SQL = `")
text = text.replace("`);\n    stmt.run(", "`;\n    await turso.execute({ sql: SQL, args: [")
text = text.replace("`);\n    stmt.get(", "`;\n    return (await turso.execute({ sql: SQL, args: [")

text = text.replace("stmt.run", "await turso.execute({ sql: SQL, args: [")

text = text.replace("db.exec(\"ALTER", "await turso.execute(\"ALTER")

# Change types
text = text.replace("Database.Database", "any")

# Now add await to internal function calls:
funcs = ["createRoom", "getRoomByCode", "joinRoom", "leaveRoom", "updateRoomDraw", "updateRoomRoster", "updateRoomCards", "updateRoomStatus", "syncHostedRoomParticipants", "getRoomById"]
for f in funcs:
    text = re.sub(r"(?<!await )" + f + r"\(", f"await {f}(", text)

with open('src/lib/room-db.ts', 'w') as f:
    f.write(text)
